### PR TITLE
Table of contents for API reference pages

### LINKED ISSUE #21808: feat(www) Table of contents for API reference pages
## Summary

API reference pages, such as [node-apis](https://www.gatsbyjs.org/docs/node-apis/) do not use our standard table of contents, instead using an inline component:

API table of contents

This is suboptimal, since it's stuck on the page content and we don't get the nice table of contents sidebar like we do in normal docs.

## Implementation

We'd need to change the [table of contents](https://github.com/gatsbyjs/gatsby/blob/master/www/src/components/docs-table-of-contents.js#L64) component to take in a generic list of items rather than a `page` objects. Once this is done, it should be fairly straightforward to add it to the [api template](https://github.com/gatsbyjs/gatsby/blob/master/www/src/templates/template-api-markdown.js).

## Related Issues

We may want to wait until #21762 to go in, since that also changes the table of contents component a fair bit.

### PR DESCRIPTION
**Pull request by andresz1**
## Description

This is my first PR! and attends to solve issue #21808 integrating the new `TableOfContent` component in `template-api-markdown`. Please let me know any feedback and I'd be more than happy to improve the PR. I noticed some repeated logic between `template-api-markdown` and `template-docs-markdown` and I was wondering if I should improve it in this PR or in a separate one.

### Documentation
- In think that no documentation needs to be updated.

## Related Issues
- Fixes #21808

## Preview
gatsby-toc

### CODE REVIEW

File: www/src/templates/template-api-markdown.js [hunk 1]
```diff
@@ -115,13 +152,46 @@ export default function APITemplate({ data, location, pageContext }) {
...
+                maxWidth: `mainContentWidth.withSidebar`,
+                minWidth: 0,
+              },
+            }}
+          >
             <MDXRenderer slug={page.fields.slug}>{page.body}</MDXRenderer>
-            <h2>{page.frontmatter.contentsHeading || "APIs"}</h2>
-            <APIContents docs={mergedFuncs} />
-            <h2>Reference</h2>
+            <h2 id="APIs">{page.frontmatter.contentsHeading || "APIs"}</h2>
```
Comment (andresz1) [inline www/src/templates/template-api-markdown.js:194]: Not sure about this. Maybe would be better to use an `Anchor` component ? 🤔

File: www/src/templates/template-api-markdown.js [hunk 2]
```diff
@@ -82,6 +89,33 @@ export default function APITemplate({ data, location, pageContext }) {
   const mergedFuncs = mergeFunctions(data, pageContext)
   const description = page.frontmatter.description || page.excerpt
 
+  const getAPIItems = () => {
+    return {
+      title: "APIs",
```
Comment (tesseralis) [inline www/src/templates/template-api-markdown.js:94]: This should be `frontmatter.contentsHeading` (if it exists).

File: www/src/templates/template-api-markdown.js [hunk 3]
```diff
@@ -82,6 +89,33 @@ export default function APITemplate({ data, location, pageContext }) {
...
+      url: '#APIs',
```
Comment (tesseralis) [inline www/src/templates/template-api-markdown.js:95]: Can you extract the name of the id to a variable so it's clear what what the ID is being used for?

File: www/src/templates/template-api-markdown.js [hunk 4]
```diff
@@ -74,6 +75,12 @@ const mergeFunctions = (data, context) => {
   return mergedFuncs
 }
 
+const containerStyles = {
```
Comment (tesseralis) [inline www/src/templates/template-api-markdown.js:78]: Most of these styles are copied from `template-docs-markdown` right? It would be helpful to refactor things to that we don't repeat the same style definitions across two files.

We could make a component called something like `<DocsMarkdownPage>` and put styles common between the two there.

File: www/src/templates/template-api-markdown.js (hunk 3 above)
Comment (andresz1) [reply www/src/templates/template-api-markdown.js:95]: Sure!

File: www/src/templates/template-api-markdown.js (hunk 4 above)
Comment (andresz1) [reply www/src/templates/template-api-markdown.js:78]: Yep I thought the same! I'll do it 😄

**Comment by andresz1** [comment]
> Thank you so much! This is going to make navigating the AP pages so much cleaner. I have a couple of suggestions for code organization and clarity. Thank you for contributing!

Thank you for the great feedback! @tesseralis I'll do the fixes and I'll let you know

**Comment by andresz1** [comment]
Hi @tesseralis,

I just performed the requested changes. I think that now is way better 🙌can you check it again when you have time please ? also I'm not sure why `gatsby-pnp` tests are failing 😞

File: www/src/components/docs-markdown-page.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    `calc(${t.sizes.mainContentWidth.withSidebar} + ${t.sizes.tocWidth} + ${t.space[9]} + ${t.space[9]} + ${t.space[9]})`,
+  px: 9,
+}
+
+function DocsMarkdownPage({
+  page,
+  location,
+  prev,
+  next,
+  items: itemsProp = [],
```
Comment (tesseralis) [inline www/src/components/docs-markdown-page.js:33]: I would rename the `items` and `depth` props to `tableOfContentsItems` and `tableOfContentsDepth` to make it clearer where these props are used.

File: www/src/components/docs-markdown-page.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  ...others
+}) {
+  const getTOCItems = () => {
+    if (page.frontmatter.disableTableOfContents) {
+      return []
+    }
+
+    const items = page.tableOfContents.items || [];
+
+    return [...items, ...itemsProp]
```
Comment (tesseralis) [inline www/src/components/docs-markdown-page.js:45]: I would put this logic in `template-api-markdown` instead. Then you can put the defaults right in the destructure:

```
function DocsMarkdownPage({
  page,
  tableOfContentsItems = page.tableOfContents.items,
  tableOfContentsDepth = page.frontmatter.tableOfContentsDepth,
})

File: www/src/components/docs-markdown-page.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  location,
+  prev,
+  next,
+  items: itemsProp = [],
+  depth: depthProp,
+  children,
+  ...others
+}) {
+  const getTOCItems = () => {
+    if (page.frontmatter.disableTableOfContents) {
```
Comment (tesseralis) [inline www/src/components/docs-markdown-page.js:39]: I think it's more clear to put the check for the `disableTableOfContents` prop when you're calculating `isTOCVisible` below.

File: www/src/templates/template-api-markdown.js [hunk 8]
```diff
@@ -75,59 +63,42 @@ const mergeFunctions = (data, context) => {
...
+      items: mergedFuncs.map(mergedFunc => ({
+        url: `#${mergedFunc.name}`,
+        title: mergedFunc.name,
+      })),
+    }]
+  }
 
   // Cleanup graphql data for usage with API rendering components
   const mergedFuncs = mergeFunctions(data, pageContext)
+  const items = getAPIItems(mergedFuncs)
```
Comment (tesseralis) [inline www/src/templates/template-api-markdown.js:84]: I would just inline this function, since it's only being called once.

**Comment by andresz1** [comment]
Thank you so much @tesseralis! I performed your requested changes and also I added some tests :3 when you have time can you check again ?

File: www/src/templates/template-api-markdown.js [hunk 9]
```diff
@@ -75,59 +64,44 @@ const mergeFunctions = (data, context) => {
...
-  const { next, prev } = pageContext
+  const { prev, next } = pageContext
   const page = data.mdx
+  const { frontmatter, tableOfContents } = page
+  const heading = frontmatter.contentsHeading || "APIs"
+  const headingId = "apis"
 
   // Cleanup graphql data for usage with API rendering components
   const mergedFuncs = mergeFunctions(data, pageContext)
+  const items = tableOfContents.items || []
```
Comment (tesseralis) [inline www/src/templates/template-api-markdown.js:75]: Could you add a comment here? Something like:

```
// Generate table of content items for API entries.

File: www/src/components/__tests__/docs-markdown-page.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  const { getByText } = setup()
+
+  expect(getByText("Table of Contents")).toBeDefined()
+})
+
+it("should not display table of content if there are no items", () => {
+  const tableOfContentsItems = []
+  const { queryByText } = setup({ tableOfContentsItems })
+
+  expect(queryByText("Table of Contents")).toBeNull()
```
Comment (tesseralis) [inline www/src/components/__tests__/docs-markdown-page.js:103]: Could we do this instead, for this and the other calls to `toBeNull`?

```suggestion
expect(queryByText("Table of Contents")).not.toBeDefined()
```

This way we don't get a false failure if one of these returnes `undefined` for some reason instead of `null`.

File: www/src/components/__tests__/docs-markdown-page.js [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  }
+})
+
+jest.mock("../../hooks/use-site-metadata", () => {
+  return () => {
+    return { siteUrl: "https://www.gatsbyjs.org" }
+  }
+})
+
+Object.defineProperty(window, "IntersectionObserver", {
```
Comment (tesseralis) [inline www/src/components/__tests__/docs-markdown-page.js:29]: Hmm, I see that this is copied from the table of contents tests. Ideally we'd do something like what is recommended [in the Jest docs](https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom) but we can leave that for a separate PR.

Comment (andresz1) [reply www/src/components/__tests__/docs-markdown-page.js:29]: Yes! I agree to create another PR for that and maybe we can also add a test-utils file or similar as shown in react-testing-library [docs](https://testing-library.com/docs/react-testing-library/setup) where we could define a custom render with all the needed providers since now is repeated in the tests. What do you think @tesseralis ?

File: www/src/components/__tests__/docs-markdown-page.js (hunk 10 above)
Comment (andresz1) [reply www/src/components/__tests__/docs-markdown-page.js:103]: I tried but it doesn't work 😞Checking the doc, [queryBy](https://testing-library.com/docs/dom-testing-library/api-queries#queryby) always return `null` if no element match. I could change it to `toBeFalsy` 🤔 but I think that with `toBeNull` in this case is just fine. Let me what you think!

Comment (tesseralis) [reply www/src/components/__tests__/docs-markdown-page.js:103]: Oh! in that case that should be fine then. :)

**Comment by tesseralis** [comment]
Oh! One more nit: could you delete the `APIContents` component in `components/api-reference` as it's no longer needed?

**Comment by tesseralis** [comment]
Going to ping @kylegill here since this changes the website docs.

**Comment by andresz1** [comment]
Thank you so much for your feedback @tesseralis @gillkyle! I had a lot fun doing this issue. I just removed `APIContents` (sorry that I forgot to remove it 😅)

**Comment by windmaomao** [comment]
To be honest, by following this conversation, i still don't exactly know how to add table of content, is there a documentation or a flag ?
